edb/edgeql/compiler/polyres.py

```diff
@@ -238,7 +240,15 @@ def _get_cast_distance(
                 resolved_poly_base_type = resolved
 
             if resolved_poly_base_type == resolved:
-                return s_types.MAX_TYPE_DISTANCE if is_abstract else 0
+                if is_abstract:
+                    return s_types.MAX_TYPE_DISTANCE
+                elif arg_type.is_range() and param_type.is_multirange():
+                    # Ranges are implicitly cast into multiranges of the same
```

Comment:
Are we confident that's the behavior we want?

Reply:
Yes, because it makes it so much easier to work with multiranges when you don't need to explicitly upcast ranges. The operations like `+`, `-`, `contains`, `overlaps`, etc. all can work with a mixture of ranges and multiranges.

The relationship between `range` and `multirange` is more like that of `int32` and `float64` than numbers and arrays. Every `int32` _can_ be represented as `float64`, but `float64` has more things it can represent, such as the result of division of 2 `int32` numbers, which _only sometimes_ is another `int32`. Similarly every `range` has a corresponding exact `multirange` to represent it, but you can do more elaborate operations on multiranges that aren't merely storing or removing "elements" like for arrays. Consider this:
```
edgedb> select range(1, 10) - range(2, 3);
edgedb error: InvalidValueError: result of range difference would not be contiguous
edgedb> select multirange([range(1,10)]) - range(2,3);
[
  [
    {"lower": 1, "upper": 2, "inc_lower": true, "inc_upper": false},
    {"lower": 3, "upper": 10, "inc_lower": true, "inc_upper": false}
  ]
]
```